File: java/leetcode/twoPointers/reverse_vowels.java

/**
 * 345. Reverse Vowels of a String
 * Write a function that takes a string as input and reverse only the vowels of a string.

Example 1:

Input: "hello"
Output: "holle"
Example 2:

Input: "leetcode"
Output: "leotcede"
Note:
The vowels does not include the letter "y".
 */

import java.util.HashSet;
import java.util.Arrays;


public class reverse_vowels {

    private final static HashSet<Character> vowels = new HashSet<>(
        Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U')
    );

    public String reverseVowels(String s) {
        if (s == null) return null;
        int i = 0, j = s.length() - 1;
        char[] result = new char[s.length()];
        while (i <= j) {
            char ci = s.charAt(i);
            char cj = s.charAt(j);
            if (!vowels.contains(ci)) {
                result[i++] = ci;
            } else if (!vowels.contains(cj)) {
                result[j--] = cj;
            } else {
                result[i++] = cj;
                result[j--] = ci;
            }
        }
        return new String(result);
    }

}

File: golang/leetcode/hashmaps/reverse_vowels.go
package hashmaps

// 345. Reverse Vowels of a String

func reverseVowels(s string) string {
	vowels := map[byte]bool{'a': true, 'e': true, 'i': true, 'o': true, 'u': true,
		'A': true, 'E': true, 'I': true, 'O': true, 'U': true}

	if len(s) == 0 {
		return s
	}
	i, j := 0, len(s)-1
	result := make([]byte, len(s))
	for i <= j {
		ci, cj := s[i], s[j]
		if _, ok := vowels[ci]; !ok {
			result[i] = ci
			i++
		} else if _, ok := vowels[cj]; !ok {
			result[j] = cj
			j--
		} else {
			result[i] = cj
			i++
			result[j] = ci
			j--
		}
	}
	return string(result)
}
